core/src/solution_submission.rs

```diff
@@ -211,27 +226,32 @@ impl<'a> StableXSolutionSubmitting for StableXSolutionSubmitter<'a> {
             });
             let cancel_future = self.cancel_transaction_after_deadline(batch_index, nonce);
 
+            // Run both futures at the same time. When one of them completes check whether the
+            // result is a "nonce already used error". If this is the case then the other future's
+            // transaction must have gone through so return that one instead.
+            // We need to handle this error because exactly one of the transactions will go through
+            // but we might observe the other transaction failing first.
             futures::pin_mut!(cancel_future);
             match future::select(submit_future, cancel_future).await {
```

Comment:
Is there a reason why you prefer this over the `futures::select!` macro? I find it a bit more ergonomic so you don't have to deal with `Either`. I think it just requires the futures to be `.shared()` so that they are fused and pinned, and then they can be reused after the `select!` invocation.

Reply:
I find function + Either more natural than a macro.  With select! I am less sure about what control flow happens internally. In this case we also need access to the future that is still running when the other one completes. It isn't clear from the select! documentation whether that is possible.

Reply:
To be clear, I fine with how it is now, I was just curious.

Reply:
Why not reuse the `select_ignoring_nonce_error` function you wrote in the other PR? With the current logic, I'd be slightly worried that there is still a chance the cancellation tx failed due to "nonce too low", yet the solution submission tx might have gone through (I know we wait 30s before we send the tx, but let there be some sync issue between the nodes or a small reorg). I think it would be better to actually wait on the result of the submit future.

Reply:
I don't understand. If the cancellation tx has nonce too low then we do wait on the submit tx.

Reply:
You are right, the logic seems correct the way it is. 
I'm still wondering why the `select_ignoring_nonce_error` would not be applicable here?

Reply:
That function is only introduced in the next PR and is specific to the retry module because of the return types and because of the fact that it takes a vector instead of working on two futures. I can see that a generalization might be possible so that the same function can be used in both cases but I'd rather do that after both PRs are done.

Reply:
Doing it in a separate PR makes sense. My reasoning for reusing it is that the logic itself here is rather complicated (cf. my misunderstanding of how the thing worked) and thus could benefit from being abstracted and tested away in a common logic part.